Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean

```diff
@@ -0,0 +1,572 @@
+/-
+Copyright (c) 2022 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+
+! This file was ported from Lean 3 source module number_theory.legendre_symbol.mul_character
+! leanprover-community/mathlib commit f0c8bf9245297a541f468be517f1bde6195105e9
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.CharP.Basic
+import Mathlib.Algebra.EuclideanDomain.Instances
+import Mathlib.Data.Fintype.Units
+
+/-!
+# Multiplicative characters of finite rings and fields
+
+Let `R` and `R'` be a commutative rings.
+A *multiplicative character* of `R` with values in `R'` is a morphism of
+monoids from the multiplicative monoid of `R` into that of `R'`
+that sends non-units to zero.
+
+We use the namespace `MulChar` for the definitions and results.
+
+## Main results
+
+We show that the multiplicative characters form a group (if `R'` is commutative);
+see `MulChar.commGroup`. We also provide an equivalence with the
+homomorphisms `Rˣ →* R'ˣ`; see `MulChar.equivToUnitHom`.
+
+We define a multiplicative character to be *quadratic* if its values
+are among `0`, `1` and `-1`, and we prove some properties of quadratic characters.
+
+Finally, we show that the sum of all values of a nontrivial multiplicative
+character vanishes; see `MulChar.IsNontrivial.sum_eq_zero`.
+
+## Tags
+
+multiplicative character
+-/
+
+
+section DefinitionAndGroup
+
+/-!
+### Definitions related to multiplicative characters
+
+Even though the intended use is when domain and target of the characters
+are commutative rings, we define them in the more general setting when
+the domain is a commutative monoid and the target is a commutative monoid
+with zero. (We need a zero in the target, since non-units are supposed
+to map to zero.)
+
+In this setting, there is an equivalence between multiplicative characters
+`R → R'` and group homomorphisms `Rˣ → R'ˣ`, and the multiplicative characters
+have a natural structure as a commutative group.
+-/
+
+
+universe u v
+
+section Defi
+
+-- The domain of our multiplicative characters
+variable (R : Type u) [CommMonoid R]
+
+-- The target
+variable (R' : Type v) [CommMonoidWithZero R']
+
+/-- Define a structure for multiplicative characters.
+A multiplicative character from a commutative monoid `R` to a commutative monoid with zero `R'`
+is a homomorphism of (multiplicative) monoids that sends non-units to zero. -/
+structure MulChar extends MonoidHom R R' where
+  /-- Require that non-units map to `0`. -/
```

Comment:
Docs added during the port.

Reply:
Hmm, the linter shouldn't complain about this, because the doc linter has been disabled in CI (for now).

Reply:
You are right, I noticed this by running the linter locally and decided to act on it, since I assumed that the CI linter error was this one.  When the CI linter was still unhappy, I realized that it was a different linting error...

Reply:
Same as above: the local linter picked up these two.